packages/core/src/auth/sso/ssoAccessTokenProvider.ts

```diff
@@ -663,9 +674,52 @@ class WebAuthorization extends SsoAccessTokenProvider {
         const cachedRegistration = await this.cache.registration.load(cacheKey)
 
         if (cachedRegistration && (isExpired(cachedRegistration) || cachedRegistration.flow !== 'web auth code')) {
-            await this.invalidate()
+            await this.invalidate('registrationExpired:WebAuth')
         }
 
         return loadOr(this.cache.registration, cacheKey, () => this.registerClient())
     }
 }
+
+/**
+ * Remembers the reason an SSO session was put in to a "needs reauthentication" state.
+ * The current use is for telemetry. When the user reauths, we want {@link AwsLoginWithBrowser}
+ * to know why it needed to be reauthed.
+ *
+ * The flow is to use `setReason()` to remember why the user was put in to a reauth state,
+ * then upon the next reauth use `getReason()`. Finally, use `clearReason()` if the reauth is
+ * successful.
+ */
+export class ReAuthReasonState {
+    private readonly memento: vscode.Memento
+    static readonly reasonNotSetState = 'reasonNotSet'
+
+    protected constructor(memento: vscode.Memento) {
+        this.memento = memento
+    }
+
+    static #instance: ReAuthReasonState
+    static get instance() {
+        return (this.#instance ??= new ReAuthReasonState(globals.context.globalState))
```

Comment:
why use globalState instead of just a variable? what scenarios are we anticipating here?

Reply:
If they are put in to a reauth state we do not know when they will reauth next. An IDE reload or extension update would wipe the state if it was in a variable.

Reply:
If the IDE reloads or restarts, is it actually valid to say that the "reauth reason" is something that happened minutes or hours ago?

Reply:
Yes, I would say so. It is the reason their connection was put in to a reauth state. Days or weeks later the cause is still the same, only once they reauth do we consider it stale.

Reply:
How common is the scenario that a vscode restart would happen after a reauth? It seems like an accident, and not connected to the reauth. Normally, the reauth will happen in the same session, so storing it in memory will capture enough data to be representative of the problems that we want to notice. 

If that's correct, then we don't need to store this in globalState, and that opens the door for using telemetry.record() as described in my other comments in this PR.

Reply:
If we are fine with assumption, then yes we can ignore storing it in globalState.

As for the implementation I will look in to something you mentioned like `telemetry.xx.recordNext()` as the telemetry context will be different and `record()` will not work as it currently is